libafl/src/observers/cmp.rs

```diff
@@ -229,65 +225,43 @@ pub trait CmpMap: Debug {
 }
 
 /// A [`CmpObserver`] observes the traced comparisons during the current execution using a [`CmpMap`]
-pub trait CmpObserver<'a, CM, S, M>: Observer<S>
-where
-    CM: CmpMap,
-    S: UsesInput,
-    M: CmpObserverMetadata<'a, CM>,
-{
+pub trait CmpObserver {
+    type Map;
+    type Metadata;
+
     /// Get the number of usable cmps (all by default)
     fn usable_count(&self) -> usize;
 
     /// Get the `CmpMap`
-    fn cmp_map(&self) -> &CM;
+    fn cmp_map(&self) -> &Self::Map;
 
     /// Get the `CmpMap` (mutable)
-    fn cmp_map_mut(&mut self) -> &mut CM;
-
-    /// Get the observer data. By default, this is the default metadata aux data, which is `()`.
-    fn cmp_observer_data(&self) -> M::Data {
-        M::Data::default()
-    }
-
-    /// Add [`struct@CmpValuesMetadata`] to the State including the logged values.
-    /// This routine does a basic loop filtering because loop index cmps are not interesting.
-    fn add_cmpvalues_meta(&mut self, state: &mut S)
-    where
-        S: HasMetadata,
-    {
-        #[allow(clippy::option_if_let_else)] // we can't mutate state in a closure
-        let meta = state.metadata_or_insert_with(|| M::new_metadata());
-
-        let usable_count = self.usable_count();
-        let cmp_observer_data = self.cmp_observer_data();
-
-        meta.add_from(usable_count, self.cmp_map_mut(), cmp_observer_data);
-    }
+    fn cmp_map_mut(&mut self) -> &mut Self::Map;
 }
 
 /// A standard [`CmpObserver`] observer
 #[derive(Serialize, Deserialize, Debug)]
 #[serde(bound = "CM: serde::de::DeserializeOwned")]
-pub struct StdCmpObserver<'a, CM, S, M>
+pub struct StdCmpObserver<'a, CM, M>
 where
-    CM: CmpMap + Serialize,
-    S: UsesInput + HasMetadata,
+    CM: Serialize,
```

Comment:
Is this constraint necessary?

Reply:
yes because this observer needs to be serialized
so cm: serialize
and m: cmpobservermetadata necessary too, because data has the type of M::Data

Reply:
Right, but the derived impl for Serialize already constrains CM, so you don't need to specify this here.

Reply:
ok added that to bound(..)